src/libraries/System.Text.RegularExpressions/src/System/Text/RegularExpressions/RegexPrefixAnalyzer.cs

```diff
@@ -11,6 +11,316 @@ namespace System.Text.RegularExpressions
+                // and if we exceed the max number of prefixes by a non-trivial amount, we'll fail the operation.
+                Span<char> setChars = stackalloc char[MaxPrefixes]; // limit how many chars we get from a set based on the max prefixes we care about
+
+                // Loop down the left side of the tree, looking for a starting node we can handle. We only loop through
+                // atomic and capture nodes, as the child is guaranteed to execute once, as well as loops with a positive
+                // minimum and thus at least one guaranteed iteration.
+                while (true)
+                {
+                    switch (node.Kind)
+                    {
+                        // These nodes are all guaranteed to execute at least once, so we can just
+                        // skip through them to their child.
+                        case RegexNodeKind.Atomic:
+                        case RegexNodeKind.Capture:
+                        case RegexNodeKind.Loop or RegexNodeKind.Lazyloop when node.M > 0:
+                            node = node.Child(0);
+                            continue;
+
+                        // Zero-width anchors and assertions don't impact a prefix and may be skipped over.
+                        case RegexNodeKind.Bol:
+                        case RegexNodeKind.Eol:
+                        case RegexNodeKind.Boundary:
+                        case RegexNodeKind.ECMABoundary:
+                        case RegexNodeKind.NonBoundary:
+                        case RegexNodeKind.NonECMABoundary:
+                        case RegexNodeKind.Beginning:
+                        case RegexNodeKind.Start:
+                        case RegexNodeKind.EndZ:
+                        case RegexNodeKind.End:
+                        case RegexNodeKind.Empty:
+                        case RegexNodeKind.UpdateBumpalong:
+                        case RegexNodeKind.PositiveLookaround:
+                        case RegexNodeKind.NegativeLookaround:
+                            return true;
+
+                        // If we hit a single character, we can just return that character.
+                        // This is only relevant for case-sensitive searches, as for case-insensitive we'd have sets for anything
+                        // that produces a different result when case-folded, or for strings composed entirely of characters that
+                        // don't participate in case conversion. Single character loops are handled the same as single characters
+                        // up to the min iteration limit. We can continue processing after them as well if they're repeaters such
+                        // that their min and max are the same.
+                        case RegexNodeKind.One or RegexNodeKind.Oneloop or RegexNodeKind.Onelazy or RegexNodeKind.Oneloopatomic when !ignoreCase || !RegexCharClass.ParticipatesInCaseConversion(node.Ch):
```

Comment:
This is one case where we may fall off Teddy if there's a non-ASCII character in the first 2/3 of any value:
https://github.com/dotnet/runtime/blob/6b26684883e4b4bad25a271c231d328f770563d5/src/libraries/System.Private.CoreLib/src/System/SearchValues/Strings/StringSearchValues.cs#L212-L222

Would it make sense to bake in such knowledge of the implementation into `FindPrefixes` and reject such sets?

---

P.S: Note to self, we may want to improve the `nonAsciiAffectedByCaseConversion` check to be closer to `RegexCharClass.ParticipatesInCaseConversion` as a follow up. https://github.com/dotnet/runtime/blob/6b26684883e4b4bad25a271c231d328f770563d5/src/libraries/System.Private.CoreLib/src/System/SearchValues/Strings/StringSearchValues.cs#L384-L387

Reply:
I'd like to avoid such special cases, at least for now. Let's see how things shake out without it, and if we hit issues, we can work to address it either in SearchValues or Regex (preferably the former).